### PR TITLE
[ContactStructuralMechanicsApplication] Replace OSUtilities with Kratos::filesystem

### PR DESCRIPTION
**Pull request by loumalouomega**
Just that, removes deprecated warnings

### CODE REVIEW

**Review by philbucher — DISMISSED** [review]
:+1:

**Review by philbucher — DISMISSED** [review]

**Comment by loumalouomega** [comment]
@philbucher any idea why does not compile?

File: applications/ContactStructuralMechanicsApplication/tests/cpp_tests/test_mixedulm_linear_solver.cpp [hunk 1]
```diff
@@ -64,15 +62,16 @@ namespace Kratos
...
+            buffer_lhs.open((Kratos::filesystem::current_path(ec)).string() + "/A_testing_condensation.mm",std::ios::out);
             std::ostream os_lhs(&buffer_lhs);
             os_lhs << lhs;
             buffer_lhs.close();
 
             const std::string rhs = "%%MatrixMarket matrix array real general\n16 1\n0\n0\n0\n-5.1725e+07\n0\n-1.0345e+08\n0\n-5.1725e+07\n0\n-1.0345e+08\n0\n-6.23386e+08\n0\n-6.23386e+08\n0\n0\n";
 
             std::filebuf buffer_rhs;
-            buffer_rhs.open(OSUtilities::GetCurrentWorkingDir() + "/b_testing_condensation.rhs",std::ios::out);
+            buffer_rhs.open((Kratos::filesystem::current_path(ec)).string() + "/b_testing_condensation.rhs",std::ios::out);
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/tests/cpp_tests/test_mixedulm_linear_solver.cpp]:
```suggestion
            buffer_rhs.open((Kratos::filesystem::current_path()).string() + "/b_testing_condensation.rhs",std::ios::out);
```

**Comment by philbucher** [comment]
not really
maybe try without the error codes?

**Comment by loumalouomega** [comment]
> not really
> maybe try without the error codes?

Nope :P

**Comment by philbucher** [comment]
hm very strange
It is the exact same code that works in the Core

Sorry for the inconvenience, I will take a closer look in the next days

**Comment by loumalouomega** [comment]
> 
> 
> hm very strange
> It is the exact same code that works in the Core
> 
> Sorry for the inconvenience, I will take a closer look in the next days

I created #6224, let's see

**Comment by philbucher** [comment]
I think it is some sort of visibility problem:
~~~
 KratosContactStructuralMechanicsCore_CXX_unity_25_48.obj : error LNK2019: unresolved external symbol "public: __cdecl ghc::filesystem::path::~path(void)" (??1path@filesystem@ghc@@QEAA@XZ) referenced in function "private: virtual void __cdecl Kratos::Testing::TestMixedULMLinearSolverRealSystem::TestFunction(void)" (?TestFunction@TestMixedULMLinearSolverRealSystem@Testing@Kratos@@EEAAXXZ) [D:\a\Kratos\Kratos\build\Release\applications\ContactStructuralMechanicsApplication\KratosContactStructuralMechanicsCore_unity.vcxproj]
~~~

strange thing that the same code works in the core

@roigcarlo do you have an idea? Sorry for bothering...

File: applications/ContactStructuralMechanicsApplication/tests/cpp_tests/test_mixedulm_linear_solver.cpp [hunk 2]
```diff
@@ -935,24 +935,24 @@ namespace Kratos
...
 
             std::filebuf buffer_lhs;
-            buffer_lhs.open((Kratos::filesystem::current_path()).string() + "/A_testing_condensation.mm",std::ios::out);
+            buffer_lhs.open(Kratos::filesystem::current_path().string() + "/A_testing_condensation.mm",std::ios::out);
```
Comment (loumalouomega) [inline applications/ContactStructuralMechanicsApplication/tests/cpp_tests/test_mixedulm_linear_solver.cpp:938]: I don't think is this, probably is that the app is not properly link with the core @roigcarlo

**Comment by roigcarlo** [comment]
I have no idea how that library works. Let me take a look but it may take a while...

**Comment by philbucher** [comment]
so apparently it does not work if used from an app
I checked, to me the export of symbols looks the same as what we do in Kratos

**Comment by loumalouomega** [comment]
Looks like the solution from https://stackoverflow.com/questions/53594731/produce-lib-of-header-only-library-that-depends-on-external-resources does not work

**Comment by philbucher** [comment]
@loumalouomega I found the problem in the [docs of the lib](https://github.com/gulrak/filesystem#using-it-as-forwarding-implementation-header):

Be aware, that it is currently not supported to hide the implementation into a Windows-DLL, as a DLL interface with C++ standard templates in interfaces is a different beast. If someone is willing to give it a try, I might integrate a PR but currently working on that myself is not a priority.

This is basically what we would need :sweat_smile: Sorry that I didn't read this careful enough.

I discussed briefly with @roigcarlo and I think we have the following options:
- Remove `Kratos::filesystem` altogether. I would really not like this because I think it is very useful and I already heard from several devs that they would like to use it.
- Figure out the original problem. It is sth similar (or even the same?) problem that we solve with adding `KRATOS_API`. I checked, the lib seems to be using (at least on first sight) the same [mechanism](https://github.com/KratosMultiphysics/Kratos/blob/master/external_libraries/ghc/filesystem.hpp#L74-L79), thats why I don't get why it doesn't work. I think this would be the best solution, which we could also contribute back such that others can profit. However I think that only @roigcarlo has the knowledge to figure this one out. Or maybe he could point us in the direction.
- Do not use the version that is separated into definition/declaration, i.e. basically include `filesystem.h` wherever we need it. The big issue with this is that this will include `windows.h` everywhere, which I would say is a no-go
- Write a wrapper like we had before with the `OSUtilities`. This way we could keep the dependency in the core, but we would need to write a wrapper fct for everything that we would want to use. For fct this would probably work, but not for classes like `filesystem::path`

@KratosMultiphysics/technical-committee I guess this is back in the discussion round :)

**Comment by loumalouomega** [comment]
> 
> 
> @loumalouomega I found the problem in the [docs of the lib](https://github.com/gulrak/filesystem#using-it-as-forwarding-implementation-header):
> 
> Be aware, that it is currently not supported to hide the implementation into a Windows-DLL, as a DLL interface with C++ standard templates in interfaces is a different beast. If someone is willing to give it a try, I might integrate a PR but currently working on that myself is not a priority.
> 
> This is basically what we would need 😅 Sorry that I didn't read this careful enough.
> 
> I discussed briefly with @roigcarlo and I think we have the following options:
> 
>     * Remove `Kratos::filesystem` altogether. I would really not like this because I think it is very useful and I already heard from several devs that they would like to use it.
> 
>     * Figure out the original problem. It is sth similar (or even the same?) problem that we solve with adding `KRATOS_API`. I checked, the lib seems to be using (at least on first sight) the same [mechanism](https://github.com/KratosMultiphysics/Kratos/blob/master/external_libraries/ghc/filesystem.hpp#L74-L79), thats why I don't get why it doesn't work. I think this would be the best solution, which we could also contribute back such that others can profit. However I think that only @roigcarlo has the knowledge to figure this one out. Or maybe he could point us in the direction.
> 
>     * Do not use the version that is separated into definition/declaration, i.e. basically include `filesystem.h` wherever we need it. The big issue with this is that this will include `windows.h` everywhere, which I would say is a no-go
> 
>     * Write a wrapper like we had before with the `OSUtilities`. This way we could keep the dependency in the core, but we would need to write a wrapper fct for everything that we would want to use. For fct this would probably work, but not for classes like `filesystem::path`
> 
> 
> @KratosMultiphysics/technical-committee I guess this is back in the discussion round :)

Another optio would be to not consider forward declaration in Windows, using the OS macros. For the moment, we can fix it later, but at least it would work

**Comment by philbucher** [comment]
> Another optio would be to not consider forward declaration in Windows, using the OS macros. For the moment, we can fix it later, but at least it would work

I don't fully understand, could you give an example pls?

**Comment by loumalouomega** [comment]
> 
> 
> > Another optio would be to not consider forward declaration in Windows, using the OS macros. For the moment, we can fix it later, but at least it would work
> 
> I don't fully understand, could you give an example pls?

Basically this: "Do not use the version that is separated into definition/declaration, i.e. basically include `filesystem.h` wherever we need it. The big issue with this is that this will include `windows.h` everywhere, which I would say is a no-go", but only for Windows, at least for the moment

**Comment by philbucher** [comment]
well `windows.h` is a problem only on windows, it does not exist in other OSs
Including it in a cpp is probably fine, but if it is in a header, then it can give some problems I think

**Comment by loumalouomega** [comment]
@philbucher I found another potential solution. http://www.lazycplusplus.com/ we can try to separate the code into a header and a source file, I assume that this respects the licensing issues

**Comment by philbucher** [comment]
@KratosMultiphysics/technical-committee discussed this and came to the following conclusion:
we won't be able to solve the issues with the visibility in Win, hence I will change it to the same concept as `Kratos::Parameters`: The library is exposed through our interface.
This will require that we have an interface function for each fct in `filesystem`, but we don't need many, hence this should be a feasible solution.
I will keep the interface close to `std::filesystem`, such that the migration will be easy once we have c++17

**Comment by loumalouomega** [comment]
I close this

**Pull request closed without merging**